SalesforceFoundation/EDA
# Affiliation mappings translations not working

# Critical Changes

# Changes
* We’ve fixed an issue with the Primary Affiliation lookup field on Contact not getting populated in languages other than English if a translated field label is in use. Users wishing to have Primary Affiliations on Contact populate in English and another language (or languages) can now use API names to ensure the fields are populated in multiple languages. To take advantage of this feature, follow the steps in [this article link.]

* We’ve also included a new error message to prevent users from creating Affiliation Mappings in EDA Settings that use an existing Account Record Type name.

# Issues Closed
Closes #320 

# New Metadata

# Deleted Metadata

# Testing Notes
Steps to Reproduce the bug:

1. Create an Account with record type “Business organization” or any other record type.
2. Create a Contact. 
3. From the Contact’s related list create an “Affiliated Account” choosing the account created on step 1 for the field “Organization” and check the the box “Primary”.
4. Go to the Contact created on step2 and make sure “Primary Business organization” is populated with the Account created on step 1 (If not visible on the detail page it should be visible on edit page).
5. *Confirmed* working as expected for English language.
6. Now, change the language from Setup →  User → User record →  Change Language to Español OR  
    click your avatar on top right → Settings → Language and Time zone → Español.
7. Repeat the step 2 through 4.
8. *Bug* : The “Primary Business organization” in this case “Organización comercial principal” is not populated with the Account.

Steps to confirm the the bug is fixed:

Scenario 1: Mappings contain Field Labels (not api names) and Language is English

1. Make sure the mappings in “Affl Mappings” custom settings for the field “Primary Affl Field” has field labels (this is default mapping when EDA is created)
2. Make sure the language is English:
    Change the language from user → User record →  Change Language to English OR  
    click your avatar on top right → Settings → Language and Time zone → English
3. Create an Account with record type “Business organization” or any other record type.
4. Create a Contact. 
5. From the Contact’s related list create an “Affiliated Account” choosing the account created on step 1 for the field “Organization” and check the the box “Primary”.
6. Go to the Contact created on step2 and make sure “Primary Business organization” is populated with the Account created on step 1 (If not visible on the detail page it should be visible on edit page).
7. *Confirmed* working as expected for English language.
8. Create an Account with record type same as step 1
9. Again, from the Contact’s related list create an “Affiliated Account” choosing the account created on step 6 for the field “Organization” and check the the box “Primary”.
10. Go to the Contact created on step2 and make sure “Primary Business organization” is updated with the Account created on step 6 (If not visible on the detail page it should be visible on edit page).
11. *Confirmed* working as expected for English language.

Scenario 2: Mappings contain Field api names (not Labels) and Language is English

1. Make sure the mappings in “Affl Mappings” custom settings for the field “Primary Affl Field” has field API names
2. Make sure the language is English:
    Change the language from user → User record →  Change Language to English OR  
    click your avatar on top right → Settings → Language and Time zone → English
3. Create an Account with record type “Business organization” or any other record type.
4. Create a Contact. 
5. From the Contact’s related list create an “Affiliated Account” choosing the account created on step 1 for the field “Organization” and check the the box “Primary”.
6. Go to the Contact created on step2 and make sure “Primary Business organization” is populated with the Account created on step 1 (If not visible on the detail page it should be visible on edit page).
7. *Confirmed* working as expected for English language.
8. Create an Account with record type same as step 1
9. Again, from the Contact’s related list create an “Affiliated Account” choosing the account created on step 6 for the field “Organization” and check the the box “Primary”.
10. Go to the Contact created on step2 and make sure “Primary Business organization” is updated with the Account created on step 6 (If not visible on the detail page it should be visible on edit page).
11. *Confirmed* working as expected for English language.
    
Scenario 3: Mappings contain Field api names (not Labels) and Language is Not English

1. Make sure the mappings in “Affl Mappings” custom settings for the field “Primary Affl Field” has field API names
2. Make sure the language is not English:
    Change the language from user → User record →  Change Language to Español OR  
    click your avatar on top right → Settings → Language and Time zone → Español
3. Create an Account with record type “Business organization” or any other record type.
4. Create a Contact. 
5. From the Contact’s related list create an “Affiliated Account” choosing the account created on step 1 for the field “Organization” and check the the box “Primary”.
6. Go to the Contact created on step2 and make sure “Primary Business organization” is populated with the Account created on step 1 (If not visible on the detail page it should be visible on edit page).
7. *Confirmed* working as expected for Spanish language.
8. Create an Account with record type same as step 1
9. Again, from the Contact’s related list create an “Affiliated Account” choosing the account created on step 6 for the field “Organization” and check the the box “Primary”.
10. Go to the Contact created on step2 and make sure “Primary Business organization” is updated with the Account created on step 6 (If not visible on the detail page it should be visible on edit page).
11. *Confirmed* working as expected for Spanish language.

Steps to test error message from Settings UI:
1) Go to EDA Settings → Affiliations → Affiliation Mappings 
2) Copy an Account Record type name from Affiliation Mappings, say "Business Organization"
3) Paste it in the "New Affiliation Mappings" section in the field "Account Record Type*" and paste "Primary_Organization__c" in "Primary Affl Field*" field and hit "Add Mapping"
4) You should see a toast - error message at the top - "We can’t map this Affiliation because it references a record type that is already referenced by another Affiliation Mapping. If you want to add this mapping so the Affiliation is translated, click Edit to modify the existing mapping. Then, specify the field API name under Contact Primary Affl Field. Do not change the name specified for Account Record Type."
5) Close the error message
6) Now, Slightly update the text in the field "Account Record Type" to make it unique say, Business Organization2" with any value in "Primary Affl Field*" field and hit "Add Mapping"
7) The mapping should reflect in the above section "Affilaition Mappings" and the newly created record should also get added to the custom settings -> Affl Mappings -> Manage.
